jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c

```diff
@@ -1687,6 +1687,248 @@ ecma_builtin_typedarray_prototype_find_index (ecma_value_t this_arg, /**< this a
   return ecma_builtin_typedarray_prototype_find_helper (this_arg, predicate, predicate_this_arg, false);
 } /* ecma_builtin_typedarray_prototype_find_index */
 
+/**
+ * Helper function to normalizing an array index
+ *
+ * The function clamps the given argument to the [0, length] range.
+ *
+ * See also:
+ *          - ECMA-262 v6, 22.1.3.3 steps 7,10,13
+ *          - ECMA-262 v6, 22.2.3.23 steps 7,10
+ * Used by:
+ *          - The %TypedArray%.prototype.copyWithin routine.
+ *          - The %TypedArray%.prototype.slice routine.
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static uint32_t
+ecma_builtin_typedarray_prototype_index_helper (uint32_t length, /**< length of the typedarray */
+                                                ecma_number_t index, /**< index */
+                                                bool is_end_index) /**< true - normalize the end index */
+{
+  uint32_t ret_value;
+
+  if (!ecma_number_is_nan (index))
+  {
+    if (ecma_number_is_zero (index))
+    {
+      ret_value = 0;
+    }
+    else if (ecma_number_is_infinity (index))
+    {
+      ret_value = ecma_number_is_negative (index) ? 0 : length;
+    }
+    else if (ecma_number_is_negative (index))
+    {
+      int32_t norm_index = (int32_t) length + ecma_number_to_int32 (index);
+      ret_value = (uint32_t) JERRY_MAX (norm_index, 0);
+    }
+    else
+    {
+      ret_value = JERRY_MIN (ecma_number_to_uint32 (index), length);
+    }
+  }
+  else
+  {
+    ret_value = is_end_index ? length : 0;
+  }
+
+  return ret_value;
+} /* ecma_builtin_typedarray_prototype_index_helper */
+
+/**
+ * Helper function to get the uint32_t value from an argument.
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static uint32_t
+ecma_builtin_get_uint32_value_from_argument (ecma_value_t arg_value, /**< this argument */
+                                             ecma_value_t ret_value, /**< base function's ret_value */
+                                             uint32_t length, /**< length of the TypedArray */
+                                             bool is_end_index) /**< true - normalize the end index */
+{
+  ecma_number_t num_var;
+  uint32_t index;
+
+  ret_value = ecma_get_number (arg_value, &num_var);
+
+  if (ECMA_IS_VALUE_ERROR (ret_value))
+  {
+    return ret_value;
+  }
+
+  index = ecma_builtin_typedarray_prototype_index_helper (length, num_var, is_end_index);
```

Comment:
Can we use `ecma_builtin_helper_array_index_normalize` instead of creating another helper?
`if (is_end_index && ecma_number_is_nan (num_var))` => 0, otherwise call `ecma_builtin_helper_array_index_normalize`.